File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder print = new StringBuilder();
    static int N;
    static long B;
    static long[][] matrix;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        B = Long.parseLong(st.nextToken());
        matrix = new long[N][N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                matrix[i][j] = Long.parseLong(st.nextToken());
            }
        }
        long[][] answer = matrixPower(B);
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                print.append(answer[i][j]).append(" ");
            }
            print.append('\n');
        }
        System.out.println(print);
    }
    static long[][] matrixProd(long[][] A, long[][] B) {
        long[][] answer = new long[N][N];
        long value;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                value = 0;
                for (int k = 0; k < N; k++) {
                    value += A[i][k] * B[k][j];
                }
                answer[i][j] = value % 1000;
            }
        }
        return answer;
    }
    static long[][] matrixPower(long B) {
        long[][] res = new long[N][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (i == j) res[i][j] = 1;
            }
        }
        long[][] answer = matrix;
        while (B > 0) {
            if (B % 2 == 1) {
                res = matrixProd(res, answer);
            }
            answer = matrixProd(answer, answer);
            B /= 2;
        }
        return res;
    }
}

File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.cc
#define MOD 1000
#define ll long long

#include <iostream>
#include <vector>

using namespace std;
ll n, m;

vector<vector<ll>> matrix_product(vector<vector<ll>> A, vector<vector<ll>> B) {
    vector<vector<ll>> answer(n, vector<ll>(n, 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            ll value = 0;
            for (int k = 0; k < n; k++) {
                value += A[i][k] * B[k][j];
            }
            answer[i][j] = value % MOD;
        }
    }
    return answer;
}

vector<vector<ll>> matrix_power(vector<vector<ll>> matrix, long long k) {
    vector<vector<ll>> res(n, vector<ll>(n));
    for (int i = 0; i < n; i++) {
        res[i][i] = 1;
    }
    while (k > 0) {
        if (k % 2 == 1) {
            res = matrix_product(res, matrix);
        }
        matrix = matrix_product(matrix, matrix);
        k /= 2;
    }
    return res;
}

int main()
{
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);

    cin >> n >> m;

    vector<vector<ll>> matrix(n, vector<ll>(n));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> matrix[i][j];
        }
    }

    matrix = matrix_power(matrix, m);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cout << matrix[i][j] << " ";
        }
        cout << '\n';
    }

    return 0;
}